File: python/tests/test_parity.py
"""Parity with the core's configuration spec, in both directions.

The core's serialised defaults are the authority: every option there must
be reachable from the surface, every surface argument must serialise to a
core option, and the parity table must be exactly what the generator
renders from the spec.
"""

from __future__ import annotations

import importlib.util
import json
from pathlib import Path
from typing import Any

import numpy as np
import pytest
from thiessen import (
    GeometryParams,
    Model,
    StructureParams,
    TermParams,
    _native,
    gaussian,
    probit,
)
from thiessen.estimators import AddiVortesRegressor

ROOT = Path(__file__).resolve().parents[2]

#: Serialised groups with no field on the stable surface.
UNEXPOSED = {"mean_params.cell", "variance_params.cell"}


def _full_term() -> TermParams:
    return TermParams(
        tessellations=8,
        k=2.0,
        lambda_c=4.0,
        geometry=GeometryParams(metric=["euclidean", "euclidean"], sigma_c=0.7),
        structure=StructureParams(omega=1.5),
    )


def _full_model() -> Model:
    return Model(
        outcome=gaussian(nu=5.0, q=0.9),
        mean_params=_full_term(),
        variance_params=_full_term(),
        burn_in=5,
        draws=6,
        thinning=2,
        prior_only=True,
    )


def _paths(tree: dict[str, Any], prefix: str = "") -> set[str]:
    paths: set[str] = set()
    for key, value in tree.items():
        path = f"{prefix}{key}"
        if isinstance(value, dict) and value:
            paths |= _paths(value, path + ".")
        else:
            paths.add(path)
    return paths


def test_every_surface_argument_is_a_core_option():
    """The core accepts the fully populated surface: no silent extras."""
    _native.validate_config(_full_model()._json())
    _native.validate_config(Model(outcome=probit(offset=0.5))._json())


def test_every_core_option_is_reachable_from_the_surface():
    core = _paths(json.loads(_native.default_config()))
    surface = _paths(json.loads(_full_model()._json()))

    missing = {path for path in core - surface if path not in UNEXPOSED}
    assert not missing, f"core options unreachable from the surface: {missing}"


def test_every_outcome_family_option_is_a_constructor_argument():
    import inspect

    constructors: dict[str, Any] = {"gaussian": gaussian, "probit": probit}
    for family in json.loads(_native.outcome_defaults()):
        (name, params) = next(iter(family.items()))
        arguments = set(inspect.signature(constructors[name]).parameters)
        assert set(params) == arguments, name


@pytest.mark.skipif(
    not (ROOT / "docs" / "parity.md").is_file(), reason="not in the source tree"
)
def test_the_parity_table_is_what_the_generator_renders():
    spec = importlib.util.spec_from_file_location(
        "parity_table", ROOT / "tools" / "parity_table.py"
    )
    assert spec is not None and spec.loader is not None
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)

    committed = (ROOT / "docs" / "parity.md").read_text()
    assert module.render() == committed, (
        "docs/parity.md drifted from the core spec; regenerate with "
        "python tools/parity_table.py --write"
    )


def test_the_factor_encoding_is_the_shared_fixture():
    """Nine rows, three levels: the encoding the R suite asserts as well."""
    n = 9
    x = np.column_stack([np.arange(n) / (n - 1), np.arange(n) % 3.0])
    y = x[:, 0] + 0.5 * x[:, 1]

    model = AddiVortesRegressor(
        categorical_features=[1],
        mean_params=TermParams(tessellations=4),
        burn_in=2,
        draws=2,
        random_state=1,
    ).fit(x, y)

    encoded = model._encoding.transform([x[:, 0], x[:, 1]])
    expected = np.column_stack(
        [
            x[:, 0],
            (x[:, 1] == 1.0).astype(np.float64),
            (x[:, 1] == 2.0).astype(np.float64),
        ]
    )
    np.testing.assert_array_equal(encoded, expected)

File: r/tests/testthat/test-parity.R
# Parity with the core's configuration spec, in both directions: every
# option of the core's serialised defaults is reachable from the surface,
# and the fully populated surface is accepted by the core, so there is no
# silent extra.

full_term <- function(tessellations) {
  term_params(
    tessellations = tessellations,
    k = 2,
    lambda_c = 4,
    geometry = geometry_params(
      metric = c("euclidean", "euclidean"), sigma_c = 0.7
    ),
    structure = structure_params(omega = 1.5)
  )
}

full_control <- function() {
  thiessen_control(
    outcome = gaussian_outcome(nu = 5, q = 0.9),
    mean_params = full_term(8),
    variance_params = full_term(4),
    general_params = general_params(
      burn_in = 5, draws = 6, thinning = 2, prior_only = TRUE
    )
  )
}

config_paths <- function(tree, prefix = "") {
  paths <- character(0)
  for (name in names(tree)) {
    path <- paste0(prefix, name)
    value <- tree[[name]]
    if (is.list(value) && length(value) > 0L && !is.null(names(value))) {
      paths <- c(paths, config_paths(value, paste0(path, ".")))
    } else {
      paths <- c(paths, path)
    }
  }
  paths
}

# Serialised groups with no field on the stable surface.
UNEXPOSED <- c("mean_params.cell", "variance_params.cell")

test_that("every surface argument is a core option", {
  expect_no_error(full_control())
  expect_no_error(thiessen_control(outcome = probit_outcome(offset = 0.5)))
})

test_that("every core option is reachable from the surface", {
  core <- config_paths(
    jsonlite::fromJSON(core_defaults(), simplifyVector = FALSE)
  )
  surface <- config_paths(
    jsonlite::fromJSON(config_json(full_control()), simplifyVector = FALSE)
  )

  missing <- setdiff(setdiff(core, surface), UNEXPOSED)
  expect_identical(missing, character(0))
})

test_that("every outcome family option is a constructor argument", {
  catalogue <- jsonlite::fromJSON(
    core_outcome_defaults(),
    simplifyVector = FALSE
  )
  for (family in catalogue) {
    kind <- names(family)[[1L]]
    constructor <- get(
      paste0(kind, "_outcome"),
      envir = asNamespace("thiessen")
    )

    expect_setequal(names(formals(constructor)), names(family[[kind]]))
    expect_setequal(
      names(unclass(constructor())), names(formals(constructor))
    )
  }
})

test_that("the factor encoding is the shared fixture", {
  # Nine rows, three levels: the encoding the Python suite asserts as well.
  n <- 9
  continuous <- (seq_len(n) - 1) / (n - 1)
  level <- factor(c("l0", "l1", "l2")[(seq_len(n) - 1) %% 3 + 1])
  frame <- data.frame(y = continuous + 0.5 * (as.integer(level) - 1),
                      a = continuous, g = level)

  fit <- thiessen(
    y ~ a + g, frame,
    thiessen_control(
      tessellations = 4,
      general_params = general_params(burn_in = 2, draws = 2)
    ),
    seed = 1
  )

  expected <- cbind(
    continuous,
    as.double(level == "l1"),
    as.double(level == "l2")
  )
  expect_identical(unname(fit$x), unname(expected))
})

test_that("the package masks nothing attached by default", {
  # `gaussian()` masked `stats::gaussian()`, so a `glm(family = gaussian)`
  # after `library(thiessen)` passed a `thiessen_outcome` to `glm` and
  # failed obscurely. `R CMD check` does not flag it.
  exports <- getNamespaceExports("thiessen")
  attached <- c("stats", "graphics", "grDevices", "utils", "datasets",
                "methods", "base")

  masked <- Reduce(
    union,
    lapply(attached, function(package) {
      intersect(exports, getNamespaceExports(package))
    }),
    character(0)
  )

  expect_identical(masked, character(0))
})
